Sources/SwiftParser/Diagnostics/ParserDiagnosticMessages.swift

```diff
@@ -131,79 +132,184 @@ public struct InvalidIdentifierError: ParserError {
-    if missingToken.text.isEmpty {
-      message += " \(missingToken.tokenKind.decomposeToRaw().rawKind.nameForDiagnostics)"
+  /// Returns a list of the tokens that are missing
+  private var missingNodesDescription: String {
+    if let codeBlock = commonParent?.as(CodeBlockSyntax.self),
+        missingNodes.contains(Syntax(codeBlock.leftBrace)),
+        missingNodes.contains(Syntax(codeBlock.rightBrace)) {
+      return "code block"
+    } else if let memberDeclBlock = commonParent?.as(MemberDeclBlockSyntax.self),
+                missingNodes.contains(Syntax(memberDeclBlock.leftBrace)),
+                missingNodes.contains(Syntax(memberDeclBlock.rightBrace)) {
+      return "member block"
+    }
+
+    enum Part {
+      case sourceText(String)
+      case tokenWithoutDefaultText(RawTokenKind)
+      case node(Syntax)
+
+      var description: String {
+        switch self {
+        case .sourceText(let content):
+          return "'\(content.trimmingWhitespaces())'"
+        case .tokenWithoutDefaultText(let tokenKind):
+          return tokenKind.nameForDiagnostics
+        case .node(let node):
+          if let parent = node.parent,
+             let childName = parent.childNameForDiagnostics(node.index) {
+            return "\(childName)"
+          } else {
+            return "\(node.nodeTypeNameForDiagnostics() ?? "syntax")"
+          }
+        }
+      }
+    }
+
+    var parts: [Part] = []
+    for missingNode in missingNodes {
+      if let missingToken = missingNode.as(TokenSyntax.self) {
+        let newPart: Part
+        let (rawKind, text) = missingToken.tokenKind.decomposeToRaw()
+        if let text = text, !text.isEmpty {
+          let presentToken = TokenSyntax(missingToken.tokenKind, presence: .present)
+          newPart = .sourceText("\(Syntax(Format().format(syntax: presentToken)))")
+        } else {
+          let newKind: TokenKind
+          if let defaultText = rawKind.defaultText {
+            newKind = TokenKind.fromRaw(kind: rawKind, text: String(syntaxText: defaultText))
+            let presentToken = TokenSyntax(newKind, presence: .present)
+            newPart = .sourceText("\(Syntax(Format().format(syntax: presentToken)))")
+          } else {
+            newPart = .tokenWithoutDefaultText(rawKind)
+          }
+        }
+
+        switch (parts.last, newPart) {
+        case (.sourceText(let previousContent), .sourceText(let newContent)):
+          parts[parts.count - 1] = .sourceText(previousContent + newContent)
+        default:
+          parts.append(newPart)
+        }
+      } else {
+        parts.append(.node(missingNode))
+      }
+    }
+    let partDescriptions = parts.map({ $0.description })
+    if partDescriptions.count > 1 {
+      return "\(partDescriptions[0..<partDescriptions.count - 1].joined(separator: ", ")) and \(partDescriptions.last!)"
     } else {
-      message += " '\(missingToken.text)'"
+      return "\(partDescriptions.joined(separator: ", "))"
```

Comment:
This is 0/1 (though would only be 0 if you remove the `!` as in my previous comment). So no need to join here.

Reply:
I changed the entire `if` to a `switch` that covers the cases `0`, `1` and `>1`